barretenberg/cpp/src/barretenberg/vm2/tracegen/tx_trace.cpp

```diff
@@ -0,0 +1,219 @@
+             { Column::tx_contract_addr, event.contract_address },
+             { Column::tx_is_static, event.is_static },
+             { Column::tx_calldata_hash, event.calldata_hash },
+             { Column::tx_reverted, event.success } };
+};
+
+std::vector<std::pair<Column, FF>> handle_append_tree_phase(const simulation::PrivateAppendTreeEvent& event,
+                                                            TransactionPhase phase)
+{
+    return {
+        { Column::tx_is_tree_insert_phase, 1 },
+        { Column::tx_leaf_value, event.leaf_value },
+
+        // Selectors
+        { Column::tx_sel_non_revertible_append_note_hash, phase == TransactionPhase::NR_NOTE_INSERTION },
+        { Column::tx_sel_non_revertible_append_nullifier, phase == TransactionPhase::NR_NULLIFIER_INSERTION },
+        { Column::tx_sel_revertible_append_note_hash, phase == TransactionPhase::R_NOTE_INSERTION },
+        { Column::tx_sel_revertible_append_nullifier, phase == TransactionPhase::R_NULLIFIER_INSERTION },
+    };
+}
+
+std::vector<std::pair<Column, FF>> handle_l2_l1_msg_phase(const simulation::PrivateEmitL2L1MessageEvent& event,
+                                                          uint32_t l2_l1_msg_counter)
+{
+
+    return {
+        { Column::tx_is_l2_l1_msg_phase, 1 },
+
+        { Column::tx_l2_l1_msg_contract_address, event.scoped_msg.contractAddress },
+        { Column::tx_l2_l1_msg_recipient, event.scoped_msg.message.recipient },
+        { Column::tx_l2_l1_msg_content, event.scoped_msg.message.content },
+        { Column::tx_l2_l1_msg_counter, l2_l1_msg_counter },
+    };
+}
+
+std::vector<std::pair<Column, FF>> handle_collect_gas_fee_phase(const simulation::CollectGasFeeEvent& event)
+{
+    return {
+        { Column::tx_is_active, 1 },
+        { Column::tx_is_collect_fee, 1 },
+
+        // TODO compute fee
+
+        { Column::tx_fee_per_da_gas, event.fee_per_da_gas },
+        { Column::tx_fee_per_l2_gas, event.fee_per_l2_gas },
+        { Column::tx_max_fee_per_da_gas, event.max_fee_per_da_gas },
+        { Column::tx_max_fee_per_l2_gas, event.max_fee_per_l2_gas },
+        { Column::tx_max_priority_fees_per_l2_gas, event.max_priority_fees_per_l2_gas },
+        { Column::tx_max_priority_fees_per_da_gas, event.max_priority_fees_per_da_gas },
+    };
+}
+
+void TxTraceBuilder::process(const simulation::EventEmitterInterface<simulation::TxEvent>::Container& events,
+                             TraceContainer& trace)
+{
+    using C = Column;
+    uint32_t row = 1; // Shifts
+
+    // A nuance of the tracegen for the tx trace is that if there are no events in a phase, we still need to emit a row
+    // for this "skipped "row. This row is needed to simplify the circuit constraints and ensure that we have continuity
+    // in the tree state propagation
+
+    // We bucket the events by phase to make it easier to detect phases with no events
+    std::array<std::vector<simulation::TxEvent>, NUM_PHASES> phase_array = {};
```

Comment:
Make these (raw) pointers to TxEvents, far cheaper assuming you don't modify the source